### ISSUE TITLE
Optimize relative datetime filter clauses

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#11837]
Related: #4043

In #10980 I added QP middleware to rewrite filters against temporal fields to generate `BETWEEN` clauses or the like and avoid temporal truncation/extraction operations that break index usages. However, this improvement only applied to filters against absolute temporal moments, e.g. "show me toucan sightings where the month is January 2020". Filters against *relative* temporal moments, e.g. "show me toucan sightings in the last month" are not currently being optimized. For example this MBQL query

```clj
;; attempted-murders is our crow sighting test dataset
(mt/dataset attempted-murders
  (mt/mbql-query attempts
    {:aggregation [[:count]]
     :filter      [:time-interval $datetime :last :month]}))
```

generates this SQL (BigQuery):

```sql
SELECT count(*) AS `count` 
FROM `attempted_murders.attempts` 
WHERE datetime_trunc(`attempted_murders.attempts`.`datetime`, month)
    = datetime_trunc(datetime_add(current_datetime(), INTERVAL -1 month), month)
```

It seems possible to instead generate something along the lines of 

```sql
SELECT count(*) AS `count` 
FROM `attempted_murders.attempts` 
WHERE `attempted_murders.attempts`.`datetime` 
BETWEEN datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 month), month) 
    AND datetime_trunc(datetime_add(current_datetime(), INTERVAL -1 month), month)
```

I could probably get this working in a day. It would be a big performance win.

Since #4043 was originally marked as P1 and #10980 didn't fully solve the entire issue I'm going to mark this as P2 and address it as soon as I get a chance.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by apclapp** [comment]
Just wanted to add my two cents that this would be a huge boost to how we use Metabase. Many of our questions are alerts are count-based to measure visitor flow-- "How many users have performed X action in the past hour/day/week. Similarly, graphs that show those counts for the past day or week suffer the performance hit.

The query time for these types of queries is heavy individually (many minutes, and longer by day as the table grows), and much too long to make use of them in any dashboards.

Looking forward to this being merged in the future, and thanks for your work on #10980.

**Comment by apclapp** [comment]
@camsaul, I was wondering if this was still intended for the 36.0 launch?

**Comment by flamber** [comment]
@apclapp No, this will likely not make it for 0.36.0

**Comment by pranas** [comment]
This has become a big issue for us once the dataset grew large enough. Looking forward to try this out. Is this gonna make it into the next point release (v0.36.4 at the time of writing)?

**Comment by camsaul** [comment]
Related perf issue: #9934

**Comment by camsaul** [comment]
Implemented by #14835

**Comment by apclapp** [comment]
Huge thanks @camsaul
